### PR TITLE
MAILET-135 Introduce mailet pipeline testing

### PR DESCRIPTION
**Pull request by aduprat**
Rebased of PRs: #405 & #413

### CODE REVIEW

File: server/container/guice/memory-guice/src/main/java/org/apache/james/modules/server/MemoryMailQueueFactory.java [hunk 1]
```diff
@@ -78,12 +79,18 @@ public void enQueue(Mail mail, long delay, TimeUnit unit) throws MailQueueExcept
...
         public void enQueue(Mail mail) throws MailQueueException {
             try {
-                mailItems.addFirst(new MemoryMailQueueItem(new MailImpl(mail)));
+                mailItems.addFirst(new MemoryMailQueueItem(cloneMail(mail)));
             } catch (MessagingException e) {
                 throw new MailQueueException("Error while copying mail " + mail.getName(), e);
             }
         }
 
+        private Mail cloneMail(Mail mail) throws MessagingException {
```
Comment (mbaechler) [inline server/container/guice/memory-guice/src/main/java/org/apache/james/modules/server/MemoryMailQueueFactory.java]: does it fix a bug ? If yes, why it's not commited with the test ?

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/CommonMailetConfigurationTest.java [hunk 2]
```diff
@@ -60,6 +78,66 @@ public void tearDown() {
...
 
     @Test
-    public void startingJamesWithCommonMailetConfigurationShouldWork() {
+    public void startingJamesWithCommonMailetConfigurationShouldWork() throws Exception {
+    }
+
+    @Test
+    public void simpleMailShouldBeSent() throws Exception {
+        jamesServer.getServerProbe().addDomain("james.org");
+        String from = "user@james.org";
```
Comment (mbaechler) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/CommonMailetConfigurationTest.java]: DEFAULT_DOMAIN ?

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/CommonMailetConfigurationTest.java [hunk 3]
```diff
@@ -60,6 +78,66 @@ public void tearDown() {
...
+            calmlyAwait.atMost(Duration.ONE_MINUTE).until(() -> messageHasBeenSent(smtpClient));
+            calmlyAwait.atMost(Duration.ONE_MINUTE).until(() -> userReceivedMessage(recipient));
+        } finally {
+            smtpClient.disconnect();
+        }
+    }
+
+    private void sendMessage(SMTPClient smtpClient, String from, String recipient) {
+        try {
+            smtpClient.helo("james.org");
```
Comment (mbaechler) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/CommonMailetConfigurationTest.java]: DEFAULT_DOMAIN

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/CommonMailetConfigurationTest.java [hunk 4]
```diff
@@ -60,6 +78,66 @@ public void tearDown() {
...
+            .contains("250 2.6.0 Message received");
+    }
+
+    private boolean userReceivedMessage(String user) {
+        IMAPClient imapClient = new IMAPClient();
+        try {
+            imapClient.connect(LOCALHOST_IP, IMAP_PORT);
+            imapClient.login(user, PASSWORD);
+            imapClient.select("INBOX");
+            imapClient.fetch("1:1", "ALL");
```
Comment (mbaechler) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/CommonMailetConfigurationTest.java]: select is enough to know if you have a new email

File: server/container/guice/memory-guice/src/main/java/org/apache/james/modules/server/MemoryMailQueueFactory.java (hunk 1 above)
Comment (aduprat) [inline server/container/guice/memory-guice/src/main/java/org/apache/james/modules/server/MemoryMailQueueFactory.java]: Will squash with the last commit.

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/CommonMailetConfigurationTest.java (hunk 4 above)
Comment (aduprat) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/CommonMailetConfigurationTest.java]: Yes, but i think it's more matching the method name.

**Comment by aduprat** [comment]
test this please

**Comment by aduprat** [comment]
test this please

**Comment by aduprat** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Pull request merged**